### PR TITLE
Add 'search' to recipe manifest and implement.

### PR DESCRIPTION
**Pull request by mmandlis**
Marked couple things that I debated with // ???

### CODE REVIEW

File: runtime/manifest-parser.peg [hunk 1]
```diff
@@ -335,6 +335,17 @@ RecipeConnection
...
   }
 
+RecipeSearch
+  = 'search' whiteSpace phrase:backquotedString eolWhiteSpace tokens:(Indent (SameIndent 'tokens' t:(whiteSpace backquotedString)+ eolWhiteSpace))?
+
+  {
+    return {
+      kind: 'search',
+      phrase,
+      tokens: optional(tokens, tokens => tokens[1][2].map(t=>t[1]), null)
```
Comment (dstockwell) [inline runtime/manifest-parser.peg]: spaces around `=>`

File: runtime/manifest.js [hunk 2]
```diff
@@ -13,6 +13,7 @@ const parser = require('./build/manifest-parser.js');
 const Recipe = require('./recipe/recipe.js');
 const ParticleSpec = require('./particle-spec.js');
 const Schema = require('./schema.js');
+var Search = require('./recipe/search.js');
```
Comment (dstockwell) [inline runtime/manifest.js]: `const` for consistency

File: runtime/recipe/recipe.js [hunk 3]
```diff
@@ -24,6 +25,9 @@ class Recipe {
     // can handle. ConnectionConstraints should be a different record
     // type to particles/views.
     this._connectionConstraints = [];
+
+    // ??? should this be an array maybe?
```
Comment (dstockwell) [inline runtime/recipe/recipe.js]: Not sure yet. Make this a TODO? I think there are possibilities when merging recipes together that we might encounter multiple search strings.

File: runtime/recipe/search.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+// This code may only be used under the BSD style license found at
+// http://polymer.github.io/LICENSE.txt
+// Code distributed by Google as part of this project is also
+// subject to an additional IP rights grant found at
+// http://polymer.github.io/PATENTS.txt
+
+var assert = require('assert');
+var util = require('./util.js');
+
+// ??? rename to SearchConstraint?
```
Comment (dstockwell) [inline runtime/recipe/search.js]: :man_shrugging: Either is fine with me. Slight preference for just `Search`.

File: runtime/recipe/search.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+class Search {
+  constructor(phrase, unresolvedTokens) {
+    assert(phrase);
+
+    this._phrase = phrase;
+    // ??? should these be sets instead? do repeated tokens matter?
```
Comment (dstockwell) [inline runtime/recipe/search.js]: Lets just keep them as arrays for now.

File: runtime/manifest-parser.peg (hunk 1 above)
Comment (mmandlis) [reply runtime/manifest-parser.peg]: done

File: runtime/manifest.js (hunk 2 above)
Comment (mmandlis) [reply runtime/manifest.js]: done

File: runtime/recipe/recipe.js (hunk 3 above)
Comment (mmandlis) [reply runtime/recipe/recipe.js]: done.
and I actually noticed that merging 2 recipes both with search string won't work well, so I tweaked the implementation (and added a test)

File: runtime/recipe/search.js (hunk 4 above)
Comment (mmandlis) [reply runtime/recipe/search.js]: sgtm

File: runtime/recipe/search.js (hunk 5 above)
Comment (mmandlis) [reply runtime/recipe/search.js]: ok

**Comment by mmandlis** [comment]
I addressed all comments, do you want to take another look, or shall i merge?

**Pull request merged**